spring-kafka/src/main/java/org/springframework/kafka/support/ExceptionMatcher.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2025-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.support;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.jspecify.annotations.Nullable;
+
+import org.springframework.util.Assert;
+
+/**
+ * Configurable exception matcher with support for identifying matching exception in
+ * nested causes. The matcher can be configured in two ways:
+ * <ol>
+ * <li>With an {@linkplain #forAllowList() allow list}, only the registered exception
+ * matches. This includes exceptions that extends from any registered exception, or
+ * implement it if that's an interface.</li>
+ * <li>With a {@linkplain #forDenyList() () deny list}, an exception match only if
+ * it isn't found amongst the registered exceptions. As for the allow list, any
+ * exception that extends or implements one of the registered exception will lead it
+ * to not match.</li>
+ * </ol>
+ * By default, only the given {@link Throwable} is inspected. To also search nested
+ * causes, {@link #traverseCauses} should be enabled.
+ *
+ * @author Stephane Nicoll
+ * @author Dave Syer
+ * @author Gary Russell
```

Comment:
`@since 4.0`: https://github.com/spring-projects/spring-framework/wiki/Code-Style#use-of-since